src/vm_optimizations.nit

```diff
@@ -0,0 +1,353 @@
+						args.add arguments[i]
+						i += 1
+					end
+					self.send(p, args)
+				else if p isa MAttribute then
+					assert recv isa MutableInstance
+					write_attribute(p, recv, arguments[i])
+					i += 1
+				else abort
+			end
+			assert i == arguments.length
+
+			return send_optimize(callsite.as(not null), [recv])
+		end
+
+		return send_optimize(callsite.as(not null), arguments)
+	end
+
+	# Try to have the most efficient implementation of the method dispatch
+	fun send_optimize(callsite: CallSite, args: Array[Instance]): nullable Instance
+	do
+		var recv = args.first
+		var mtype = recv.mtype
+		var ret = send_commons(callsite.mproperty, args, mtype)
+		if ret != null then return ret
+
+		if callsite.status == 0 then callsite.optimize(recv)
+
+		var propdef
+		if callsite.status == 1 then
+			propdef = method_dispatch_sst(recv.vtable.internal_vtable, callsite.offset)
+		else
+			propdef = method_dispatch_ph(recv.vtable.internal_vtable, recv.vtable.mask,
+				callsite.id, callsite.offset)
+		end
+
+		return self.call(propdef, args)
+	end
+end
+
+redef class AAttrExpr
+	# The relative position of this attribute if perfect hashing is used,
+	# The absolute position of this attribute if SST is used
+	var offset: Int
+
+	# 0: default value
+	# 1: SST (direct access) can be used
+	# 2: PH (multiple inheritance implementation) must be used
+	var status: Int = 0
+
+	# Identifier of the class which introduced the attribute
+	var id: Int
+
+	redef fun expr(v)
+	do
+		# TODO : a workaround for now
+		if not v isa VirtualMachine then return super
+
+		var recv = v.expr(self.n_expr)
+		if recv == null then return null
+		if recv.mtype isa MNullType then fatal(v, "Receiver is null")
+		var mproperty = self.mproperty.as(not null)
+
+		assert recv isa MutableInstance
+		if status == 0 then optimize(mproperty, recv)
+
+		var i: Instance
+		if status == 1 then
+			# SST
+			i = v.read_attribute_sst(recv.internal_attributes, offset)
+		else
+			# PH
+			i = v.read_attribute_ph(recv.internal_attributes, recv.vtable.internal_vtable, recv.vtable.mask, id, offset)
+		end
+
+		# If we get a `MInit` value, throw an error
+		if i == v.initialization_value then
+			v.fatal("Uninitialized attribute {mproperty.name}")
+			abort
+		end
+
+		return i
+	end
+
+	private fun optimize(mproperty: MAttribute, recv: MutableInstance)
+	do
+		if mproperty.intro_mclassdef.mclass.positions_attributes[recv.mtype.as(MClassType).mclass] != -1 then
+			# if this attribute class has an unique position for this receiver, then use direct access
+			offset = mproperty.absolute_offset
+			status = 1
+		else
+			# Otherwise, perfect hashing must be used
+			id = mproperty.intro_mclassdef.mclass.vtable.id
+			offset = mproperty.offset
+			status = 2
+		end
+	end
+end
+
+redef class AAttrAssignExpr
+	# The relative position of this attribute if perfect hashing is used,
+	# The absolute position of this attribute if SST is used
+	var offset: Int
+
+	# 0: default value
```

Comment:
Looks like a copy paste. Maybe you can factorize this code under `AAttrFormExpr`?